src/ioctls/vcpu.rs

```diff
@@ -2375,6 +2531,129 @@ mod tests {
+        ];
+
+        for reg in &sync_regs {
+            vcpu.set_sync_dirty_reg(*reg);
+            assert_eq!(vcpu.kvm_run_ptr.as_mut_ref().kvm_dirty_regs, *reg as u64);
+            vcpu.clear_sync_dirty_reg(*reg);
+            assert_eq!(vcpu.kvm_run_ptr.as_mut_ref().kvm_dirty_regs, 0);
+        }
+
+        // Test that multiple dirty SyncRegs can be set at the same time
+        vcpu.set_sync_dirty_reg(SyncReg::Register);
+        vcpu.set_sync_dirty_reg(SyncReg::SystemRegister);
+        vcpu.set_sync_dirty_reg(SyncReg::VcpuEvents);
+        assert_eq!(
+            vcpu.kvm_run_ptr.as_mut_ref().kvm_dirty_regs,
+            SyncReg::Register as u64 | SyncReg::SystemRegister as u64 | SyncReg::VcpuEvents as u64
+        );
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    #[test]
+    fn test_sync_regs_with_run() {
+        use std::io::Write;
+
+        let kvm = Kvm::new().unwrap();
+        let vm = kvm.create_vm().unwrap();
+        // This example is based on https://lwn.net/Articles/658511/
+        #[rustfmt::skip]
+        let code = [
+            0xff, 0xc0, /* inc eax */
+            0xf4, /* hlt */
+        ];
+
+        let mem_size = 0x4000;
+        let load_addr = mmap_anonymous(mem_size);
+        let guest_addr: u64 = 0x1000;
+        let slot: u32 = 0;
+        let mem_region = kvm_userspace_memory_region {
+            slot,
+            guest_phys_addr: guest_addr,
+            memory_size: mem_size as u64,
+            userspace_addr: load_addr as u64,
+            flags: KVM_MEM_LOG_DIRTY_PAGES,
+        };
+        unsafe {
+            vm.set_user_memory_region(mem_region).unwrap();
+        }
+
+        unsafe {
+            // Get a mutable slice of `mem_size` from `load_addr`.
+            // This is safe because we mapped it before.
+            let mut slice = std::slice::from_raw_parts_mut(load_addr, mem_size);
+            slice.write_all(&code).unwrap();
+        }
+
+        let mut vcpu = vm.create_vcpu(0).unwrap();
+
+        let orig_sregs = vcpu.get_sregs().unwrap();
+
+        let mut sync_regs = vcpu.sync_regs_mut();
+
+        // Initialize the sregs in sync_regs to be the original sregs
+        sync_regs.sregs = orig_sregs;
+        sync_regs.sregs.cs.base = 0;
+        sync_regs.sregs.cs.selector = 0;
+
+        // Set up the guest to attempt to `inc rax`
+        sync_regs.regs.rip = guest_addr;
+        sync_regs.regs.rax = 0x8000;
+        sync_regs.regs.rflags = 2;
+
+        // Initialize the sync_reg flags
+        vcpu.set_sync_valid_reg(SyncReg::Register);
+        vcpu.set_sync_valid_reg(SyncReg::SystemRegister);
+        vcpu.set_sync_valid_reg(SyncReg::VcpuEvents);
+        vcpu.set_sync_dirty_reg(SyncReg::Register);
+        vcpu.set_sync_dirty_reg(SyncReg::SystemRegister);
+        vcpu.set_sync_dirty_reg(SyncReg::VcpuEvents);
+
+        // hlt is the only expected return from guest execution
+        assert!(matches!(vcpu.run().expect("run failed"), VcpuExit::Hlt));
```

Comment:
When running this test on my laptop, this is failing with:
`thread 'ioctls::vcpu::tests::test_sync_regs_with_run' panicked at 'assertion failed: matches!(vcpu.run().expect(\"run failed\"), VcpuExit :: Hlt)', src/ioctls/vcpu.rs:2648:9`
Is there something else (a capability, etc) we need to check before doing this assert?

Reply:
Ah, there is the `KVM_CAP_SYNC_REGS` capability, containing which sync_regs are available. We can add the assert in the unit tests with something like

```
let sync_regs_cap = unsafe {
        unsafe { ioctl_with_val(&vm, KVM_CHECK_EXTENSION(), Cap::SyncRegs as c_ulong) }
};
assert_eq!(sync_regs_cap, 7);
```
or opening up `check_extension_int` to `pub` rather than private.

Either way, I think we'll also need to add this to the example code to make sure the user knows to check the capability.

Any thoughts on opening `check_extension_int` to `pub`? I've never seen it,  but supposedly that capability could come back with fewer than 3 bits set, signifying that one of the three `sync_regs` isn't in use. Opening up `check_extension_int` would give the user the ability to look specifically which bits are not set rather than an all or nothing, but I'm not sure opening that up function to this one use case is worthwhile.

I've pushed this test case up to see if your capability isn't 7 to see if that was the problem.

Reply:
Yes, that was the problem. We have the `check_extension` method that is public and can be used for that, i.e.:

`if kvm.check_extension(Cap::SyncRegs) {
...
}`

And yes we should also update the documentation examples. Thanks!